packages/mdc-top-app-bar/standard/foundation.js

```diff
@@ -0,0 +1,202 @@
+/**
+ * @license
+ * Copyright 2018 Google Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import MDCTopAppBarAdapter from '../adapter';
+import MDCTopAppBarBaseFoundation from '../foundation';
+import {numbers} from '../constants';
+
+/**
+ * @extends {MDCTopAppBarBaseFoundation<!MDCTopAppBarFoundation>}
+ * @final
+ */
+class MDCTopAppBarFoundation extends MDCTopAppBarBaseFoundation {
+  /**
+   * @param {!MDCTopAppBarAdapter} adapter
+   */
+  constructor(adapter) {
+    super(adapter);
+    /**
+     * Used for diffs of current scroll position vs previous scroll position
+     * @private {number}
+     */
+    this.lastScrollPosition_ = this.adapter_.getViewportScrollY();
+
+    /**
+     * Used to verify when the top app bar is completely showing or completely hidden
+     * @private {number}
+     */
+    this.topAppBarHeight_ = this.adapter_.getTopAppBarHeight();
+
+    /**
+     * wasDocked_ is used to indicate if the top app bar was docked in the previous
+     * scroll handler iteration.
+     * @private {boolean}
+     */
+    this.wasDocked_ = true;
+
+    /**
+     * Variable for current scroll position of the top app bar
+     * @private {number}
+     */
+    this.currentAppBarScrollPosition_ = 0;
+
+    /**
+     * Used to prevent the top app bar from being scrolled out of view during resize events
+     * @private {boolean} */
+    this.isCurrentlyBeingResized_ = false;
+
+    /**
+     * The timeout that's used to throttle the resize events
+     * @private {number}
+     */
+    this.resizeThrottleId_ = -1;
+
+    /**
+     * The timeout that's used to debounce toggling the isCurrentlyBeingResized_ variable after a resize
+     * @private {number}
+     */
+    this.resizeDebounceId_ = -1;
+
+    this.scrollHandler_ = () => this.topAppBarScrollHandler_();
+    this.resizeHandler_ = () => this.topAppBarResizeHandler_();
+  }
+
+  init() {
+    super.init();
+    this.adapter_.registerScrollHandler(this.scrollHandler_);
+    this.adapter_.registerResizeHandler(this.resizeHandler_);
+  }
+
+  destroy() {
+    super.destroy();
+    this.adapter_.deregisterScrollHandler(this.scrollHandler_);
+    this.adapter_.deregisterResizeHandler(this.resizeHandler_);
+    this.adapter_.setStyle('top', '');
+  }
+
+  /**
+   * Function to determine if the DOM needs to update.
+   * @return {boolean}
+   * @private
+   */
+  checkForUpdate_() {
```

Comment:
Some of the logic is a little hard to parse mentally cuz it's so abstract. I'm wondering if adding some local variables might make it easier to read and grok.

E.g.:

```js
checkForUpdate_() {
  const offscreenBoundaryTop = -this.topAppBarHeight_;
  const hasAnyPixelsOffscreen = this.currentAppBarScrollPosition_ < 0;
  const hasAnyPixelsOnscreen = this.currentAppBarScrollPosition_ > offscreenBoundaryTop;
  const partiallyShowing = hasAnyPixelsOffscreen && hasAnyPixelsOnscreen;

  // If it was previously already docked but now is partially showing, it's no longer docked.
  if (this.wasDocked_ && partiallyShowing) {
    this.wasDocked_ = false;
  }
  // If it's not previously docked and not partially showing, it just became docked.
  else if (!this.wasDocked_ && !partiallyShowing) {
    this.wasDocked_ = true;
  }

  return partiallyShowing;
}
```